examples/preprocessing/plot_pace.py

```diff
@@ -0,0 +1,372 @@
+# points. This mirrors the sparse and irregular setting often found in
+# real-world longitudinal data.
+random_state = 12
+
+irregular_fd = irregular_sample(
+    fd,
+    8,
+    random_state=random_state,
+)
+irregular_fd.coordinate_names = fd.coordinate_names
+irregular_fd.argument_names = fd.argument_names
+irregular_fd.dataset_name = fd.dataset_name
+
+irregular_fd[:5].plot()
+plt.show()
+
+
+# %%
+# We continue by plotting all raw (t_i, X_i) points from the irregularly
+# sampled dataset. This gives insight into the data sparsity and non-uniform
+# time coverage across individuals.
+plt.figure()
+for t, v in zip(irregular_fd.points, irregular_fd.values, strict=True):
+    t_i = np.asarray(t)
+    v_i = np.asarray(v)
+    plt.scatter(t_i, v_i, alpha=0.7, s=10, color="black")
+
+plt.xlabel("age")
+plt.ylabel("height")
+plt.title("Berkeley Growth Study", pad=20)
+# plt.tight_layout()
+plt.show()
+
+# %%
+# We apply classical FPCA (intended for regularly sampled data) to the original
+# dense dataset. This gives us a reference point to later compare with PACE.
+n_components = 2
+fpca = FPCA(n_components=n_components)
+scores = fpca.fit_transform(fd)
+fpca_rec = fpca.inverse_transform(scores)
+fig, ax = plt.subplots()
+for i, sample in enumerate(fpca.components_):
+    label = fpca.components_.sample_names[i]
+    sample.plot(axes=ax, label=label)
+
+ax.set_ylim(-0.4, 0.6)
+ax.set_xlabel(fpca.components_.argument_names[0] or "Domain")
+ax.set_ylabel(fpca.components_.coordinate_names[0] or "Value")
+plt.show()
+
+fpca_rec.plot()
+plt.show()
+print(fpca.explained_variance_ratio_)
+
+# %%
+# We now apply the PACE algorithm to the irregularly sampled dataset. The
+# bandwidth parameters for mean and covariance estimation are specified
+# manually.
+pace = PACE(
+    n_components=n_components,
+    bandwidth_mean=np.array([0.1, 50]),
+    bandwidth_cov=1.5,
+)
+pace.fit(irregular_fd)
+
+fpc_scores = pace.transform(irregular_fd)
+
+# %%
+# Let's further understand the correlation of the data, by plotting the
+# covariance function, which measures how the data varies together over time.
+# First, we inspect the time point pairs by subject: although the data per
+# subject is sparse, the assembled data fill the domain of the covariance
+# surface quite densely. This visualization is relevant to justify the
+# effectiveness of the kernel smoother.
+t_mean = np.asarray(pace.mean_.grid_points[0])
+pair_counts: defaultdict[tuple[int, int], int] = defaultdict(int)
+
+for i, start in enumerate(irregular_fd.start_indices):
+    end = (
+        irregular_fd.start_indices[i + 1]
+        if i + 1 < len(irregular_fd.start_indices)
+        else len(irregular_fd.points)
+    )
+    t_i = np.asarray(irregular_fd.points[start:end])
+
+    for s in t_i:
+        for t in t_i:
+            idx_s = np.argmin(np.abs(t_mean - s))
+            idx_t = np.argmin(np.abs(t_mean - t))
+            pair_counts[(int(idx_s), int(idx_t))] += 1
```

Comment:
This is overly complicated and difficult to understand. I think you can get the values you need just by doing `np.unique(irregular_fd.points, axis=0, return_counts=True)`.